cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/RawMetricValues.java

```diff
@@ -0,0 +1,351 @@
+              aggValuesForMetric.set(resultIndex, (float) (total / counts));
+              break;
+            case MAX:
+            case LATEST:
+              aggValuesForMetric.set(resultIndex, _counts[idx] > 0 ? (float) (total / 3) : (float) (total / 2));
+              break;
+            default:
+              throw new IllegalStateException("Should never be here.");
+          }
+        // Neighbor not available, use the insufficient samples.
+        } else if (_counts[idx] > 0) {
+          aggValuesForMetric.set(resultIndex, getValue(info, idx, values));
+          if (info.id() == 0) {
+            imputations.put(resultIndex, Imputation.FORCED_INSUFFICIENT);
+          }
+        // Nothing is available, just return all 0 and NO_VALID_IMPUTATION.
+        } else {
+          aggValuesForMetric.set(resultIndex, 0);
+          if (info.id() == 0) {
+            imputations.put(resultIndex, Imputation.NO_VALID_IMPUTATION);
+          }
+        }
+        resultIndex++;
+      }
+    }
+    return new ValuesAndImputations(new AggregatedMetricValues(aggValues), imputations);
+  }
+
+  public synchronized int numSamples() {
+    int count = 0;
+    for (int i : _counts) {
+      count += i;
+    }
+    return count;
+  }
+
+  private float getValue(MetricInfo info, int index, float[] values) {
+    if (_counts[index] == 0) {
+      return 0;
+    }
+    switch (info.strategy()) {
+      case AVG:
+        return values[index] / _counts[index];
+      case MAX:
+      case LATEST:
+        return values[index];
+      default:
+        throw new IllegalStateException("Should never be here.");
+    }
+  }
+
+  private void updateValue(double newValue, MetricInfo info, int index) {
+    switch (info.strategy()) {
+      case AVG:
+        add(newValue, info.id(), index);
+        break;
+      case MAX:
+        max(newValue, info.id(), index);
+        break;
+      case LATEST:
+        latest(newValue, info.id(), index);
+        break;
+      default:
+        throw new IllegalStateException("Should never be here");
+    }
+  }
+
+  private void add(double newValue, int metricId, int index) {
+    _values.get(metricId)[index] = (float) (_counts[index] == 0 ? newValue : _values.get(metricId)[index] + newValue);
+  }
+
+  private void max(double newValue, int metricId, int index) {
+    _values.get(metricId)[index] = (float) (_counts[index] == 0 ? newValue : Math.max(_values.get(metricId)[index], newValue));
+  }
+
+  private void latest(double newValue, int metricId, int index) {
+    _values.get(metricId)[index] = (float) newValue;
+  }
+
+  private void updateValidityAndImputation(int index) {
+    if (!updateEnoughSamples(index)) {
+      if (!_imputations.get(index) && !updateAvailableAvg(index)) {
+        if (!updateAvgAdjacent(index)) {
+          if (!updateForcedInsufficient(index)) {
+            _validity.clear(index);
+            _imputations.clear(index);
```

Comment:
If there is no sample in a window, why do we clear the `_imputations`? Does it matter if we set it for this case?

Reply:
This is for clarity. None of validity and imputations means it is not valid and no imputation was used. Otherwise it might be confusing because it is not valid, but some imputation was used?